File: c#/LeetCodeProblems/LeetCodeProblems.LeetCode75/LC75_ArrayProblems.cs
using System.Text;

namespace LeetCodeProblems.LeetCode75;

public static class LC75_ArrayProblems
{
    #region Problem 238. Product of Array Except Self
    public static int[] Problem_238_ProductExceptSelf(int[] nums)
    {
        var n = nums.Length;
        var result = new int[n];
        result[0] = 1;

        for (var i = 1; i < n; i++)
        {
            result[i] = result[i - 1] * nums[i - 1];
        }

        var right = 1;
        for (int i = n - 2; i >= 0; i--)
        {
            result[i] *= right * nums[i + 1];
            right *= nums[i + 1];
        }

        return result;
    }

    #endregion

    #region Problem 151. Reverse Words in a String
    public static string Problem_151_ReverseWords(string s)
    {
        return string.Join(' ', s
            .Split(' ', StringSplitOptions.RemoveEmptyEntries)
            .Reverse());
    }
    #endregion

    #region Problem 345. Reverse Vowels of a String
    public static string Problem_345_ReverseVowels(string s)
    {
        var vowels = new char[] { 'a', 'e', 'i', 'o', 'u', 'A', 'E', 'I', 'O', 'U' };
        var left = 0;
        var right = s.Length - 1;

        var result = new StringBuilder(s);

        while (left < right)
        {
            var leftCharIsContains = vowels.Contains(result[left]);
            var rightCharIsContains = vowels.Contains(result[right]);
            
            if (leftCharIsContains && rightCharIsContains)
            {
                (result[right], result[left]) = (result[left], result[right]);
                
                right--;
                left++;
                continue;
            }

            if (!rightCharIsContains)
            {
                right--;
            }


            if (!leftCharIsContains)
            {
                left++;
            }
        }

        return result.ToString();
    }
    #endregion

    #region Problem 605. Can Place Flowers
    public static bool Problem_605_CanPlaceFlowers(int[] flowerbed, int n)
    {
        if (flowerbed[0] == 0)
        {
            flowerbed[0] = 2;
            n--;
        }

        for (int i = 1; i < flowerbed.Length; i++)
        {
            if (flowerbed[i] == 1)
            {
                if (flowerbed[i - 1] == 2)
                {
                    flowerbed[i - 1] = 0;
                    n++;
                }
            } 
            else if (flowerbed[i - 1] == 0)
            {
                flowerbed[i] = 2;
                n--;
            }
        }


        return n <= 0;
    }
    #endregion

    #region Problem 1431. Kids With the Greatest Number of Candies
    public static IList<bool> Problem_1431_KidsWithCandies(int[] candies, int extraCandies)
    {
        var max = candies.Max();
        var result = new List<bool>(candies.Length);

        for (int i = 0; i < candies.Length; i++)
        {
            if (candies[i] + extraCandies >= max)
            {
                result.Add(true);
            }
            else
            {
                result.Add(false);
            }
        }

        return result;
    }
    #endregion

    #region Problem 1071. Greatest Common Divisor of Strings
    public static string Problem_1071_GcdOfStrings(string str1, string str2)
    {
        var count = Math.Min(str1.Length, str2.Length);
        var concatString = str1 + str2;
        var result = string.Empty;

        while (count > 0)
        {
            if (str1[0..count] == str2[0..count]
                && str1.Length % count == 0
                && str2.Length % count == 0)
            {
                var substring = concatString[0..count];
                var substringsCount = concatString.Length / count;
                var isValid = true;

                for (int i = 0; i < substringsCount; i++)
                {
                    var idx = i * count;
                    if (concatString[idx..(idx + count)] != substring)
                    {
                        isValid = false;
                        break;
                    }
                }

                if (isValid)
                {
                    result = concatString[0..count];
                    break;
                }
            }

            count--;
        }

        return result;
    }
    #endregion

    #region Problem 1768. Merge Strings Alternately
    public static string Problem_1768_MergeAlternately(string word1, string word2)
    {
        var minCount = Math.Min(word1.Length, word2.Length);
        var result = new StringBuilder();

        for (int i = 0; i < minCount; i++)
        {
            result.Append(word1[i]);
            result.Append(word2[i]);
        }

        while (word1.Length > minCount)
        {
            result.Append(word1[minCount]);
            minCount++;
        }

        while (word2.Length > minCount)
        {
            result.Append(word2[minCount]);
            minCount++;
        }

        return result.ToString();
    }
    #endregion
}

File: cpp/leetCode75/lc75_arrayProblems.cpp
#include "lc75_arrayProblems.h"



std::string problem_151_reverseWords(std::string s) {
    std::stringstream ss(s);
    std::string word;

    s.clear();

    while (ss >> word)
    {
        s = word + " " + s;
    }

    
    return s.substr(0, s.length() - 1);   
}

std::string problem_345_reverseVowels(std::string s) {
    std::vector<char> vowels = { 'A', 'a', 'E', 'e', 'I', 'i', 'O', 'o', 'U', 'u' };
    std::vector<char> wordVowels;

    auto size = s.length();
    for (size_t i = 0; i < size; i++)
    {
        if (std::find(vowels.begin(), vowels.end(), s[i]) != vowels.end()) {
            wordVowels.push_back(s[i]);
        }
    }
    
    std::reverse(wordVowels.begin(), wordVowels.end());
    int idx = 0;
    for (size_t i = 0; i < size, idx < wordVowels.size(); i++)
    {
        if (std::find(vowels.begin(), vowels.end(), s[i]) != vowels.end()) {
            s[i] = wordVowels[idx];
            idx++;
        }
    }
    
    return s;
}

bool problem_605_canPlaceFlowers(std::vector<int>& flowerbed, int n) {
    if (flowerbed[0] == 0) {
        flowerbed[0] = 2;
        n--;
    }

    for (size_t i = 1; i < flowerbed.size(); i++) {
        if (flowerbed[i] == 0) {
            if (flowerbed[i - 1] == 0) {
                flowerbed[i] = 2;
                n--;
            }
        } else {
            if (flowerbed[i - 1] == 2) {
                flowerbed[i - 1] = 0;
                n++;
            }
        }
    }

    return n <= 0;       
}

std::vector<bool> problem_1431_kidsWithCandies(std::vector<int>& candies, int extraCandies) {
    auto max = *std::max_element(candies.begin(), candies.end());
    std::vector<bool> result;

    for (size_t i = 0; i < candies.size(); i++) {
        candies[i] += extraCandies;
        result.push_back(candies[i] >= max);
    }
    
    return result;
}

std::string problem_1071_gcdOfStrings(std::string str1, std::string str2) {
    auto length = std::min(str1.length(), str2.length());

    while (length > 0) {
        if (str1.length() % length == 0 && str2.length() % length == 0) {
            auto substr = str1.substr(0, length);
            auto count = str1.length() / substr.length();
            bool isValid = true;

            for (size_t i = 0; i < count; i++) {
                if (str1.substr(i * length, length) != substr) { 
                    isValid = false;
                    break; 
                }
            }
            
            if (isValid) {
                count = str2.length() / substr.length();
                for (size_t i = 0; i < count; i++) {
                    if (str2.substr(i * length, length) != substr) { 
                        isValid = false;
                        break; 
                    }
                }
            }

            if (isValid) { break; }
        }
        
        length--;
    }

    return length > 0 ? str1.substr(0, length) : std::string();
}

std::string problem_1768_mergeAlternately(std::string word1, std::string word2) {
    std::string result;
    result.reserve(word1.length() + word2.length());

    auto minCount = std::min(word1.length(), word2.length());
    auto idx = 0;
    while (idx < minCount)
    {
        result += word1[idx];
        result += word2[idx];
        idx++;
    }

    while (minCount < word1.length()) {
        result += word1[minCount];
        minCount++;
    }

    while (minCount < word2.length()) {
        result += word2[minCount];
        minCount++;
    }

    return result;
}